MapStore: Map init blocked when new nodes join during data loading

Hello,

We discovered an issue in Hazelcast 3.8.5 when using MapStores.
It seems that hcInstance.getMap(mapName) gets blocked when a new node joins and data is being loaded.
For us this is a serious problem since it might prevent the system from starting.

Here is a test that reproduces the issue:
```java
package test;

import com.hazelcast.config.Config;
import com.hazelcast.config.MapConfig;
import com.hazelcast.config.MapStoreConfig;
import com.hazelcast.config.XmlConfigBuilder;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IMap;
import org.apache.log4j.BasicConfigurator;
import org.apache.log4j.Logger;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.Assert.fail;

/**
 * Test to ensure there is no deadlock when a new node joins during data loading.
 */
public class TestMapStore33 {

    private static final Logger logger = Logger.getLogger(TestMapStore33.class);

    private static final int preloadSize = 1000;

    private static final int writeDelaySeconds = 5;

    private static final AtomicInteger mapSize = new AtomicInteger(-1);

    private final String mapName = "map" + getClass().getSimpleName();

    private final InMemoryMapStoreSleepAndCount store = new InMemoryMapStoreSleepAndCount(300);

    private final ExecutorService executorService = Executors.newFixedThreadPool(3);

    private final List<HazelcastInstance> hazelcastInstances = new ArrayList<>();

    @Before
    public void setUp() throws Exception {
        BasicConfigurator.configure();
    }

    @After
    public void tearDown() {
        for (HazelcastInstance hazelcastInstance : hazelcastInstances) {
            hazelcastInstance.getLifecycleService().terminate();
        }
    }

    @Test
    public void testMapStoreLoad() throws InterruptedException {

        // load data in the store
        store.preload(preloadSize);

        // runnable that creates a new Hazelcast instance and checks the map size
        Runnable task = new Runnable() {
            @Override
            public void run() {
                HazelcastInstance hcInstance = Hazelcast.newHazelcastInstance(getConfig());
                hazelcastInstances.add(hcInstance);
                IMap<String, String> map = hcInstance.getMap(mapName);
                int size = map.size();
                mapSize.set(size);
            }
        };

        // start first Hazelcast instance
        executorService.submit(task);

        // wait 6s so that the first instance triggers data loading on the map
        Thread.sleep(6000);

        // start second Hazelcast instance
        executorService.submit(task);

        // wait at most 150s
        final long t0 = System.currentTimeMillis();
        while ((System.currentTimeMillis() - t0) < 150000) {
            if (mapSize.get() == preloadSize) {
                break;
            }
            Thread.sleep(1000);
            logger.info(String.format("Loaded keys: %s", store.getCountLoadedKeys()));
        }

        // check for errors
        if (mapSize.get() != preloadSize) {
            fail(String.format("Not all data loaded (%s != %s).", mapSize.get(), preloadSize));
        }
    }

    private Config getConfig() {

        // create shared hazelcast config
        final Config config = new XmlConfigBuilder().build();
        config.setProperty("hazelcast.logging.type", "log4j");

        // disable JMX to make sure lazy loading works asynchronously
        config.setProperty("hazelcast.jmx", "false");

        // get map config
        MapConfig mapConfig = config.getMapConfig(mapName);

        // configure map store
        MapStoreConfig mapStoreConfig = new MapStoreConfig();
        mapStoreConfig.setEnabled(true);
        mapStoreConfig.setInitialLoadMode(MapStoreConfig.InitialLoadMode.EAGER);
        mapStoreConfig.setWriteDelaySeconds(writeDelaySeconds);
        mapStoreConfig.setClassName(null);
        mapStoreConfig.setImplementation(store);
        mapConfig.setMapStoreConfig(mapStoreConfig);

        return config;
    }

}

package test;

import com.hazelcast.core.MapStore;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

public class InMemoryMapStoreSleepAndCount implements MapStore<String, String> {

    private final ConcurrentHashMap<String, String> store = new ConcurrentHashMap<>();

    private final int msPerLoad;

    private final AtomicInteger countLoadedKeys = new AtomicInteger(0);

    // ----------------------------------------------------------- construction

    public InMemoryMapStoreSleepAndCount(int msPerLoad) {
        this.msPerLoad = msPerLoad;
    }

    public void preload(int size) {
        for (int i = 0; i < size; i++) {
            store.put("k" + i, "v" + i);
        }
    }

    // ---------------------------------------------------------------- getters

    public int getCountLoadedKeys() {
        return countLoadedKeys.get();
    }

    // ----------------------------------------------------- MapStore interface

    @Override
    public String load(String key) {
        if (msPerLoad > 0) {
            try {
                Thread.sleep(msPerLoad);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        countLoadedKeys.incrementAndGet();
        return store.get(key);
    }

    @Override
    public Map<String, String> loadAll(Collection<String> keys) {
        List<String> keysList = new ArrayList<>(keys);
        Collections.sort(keysList);

        Map<String, String> result = new HashMap<>();
        for (String key : keys) {

            if (msPerLoad > 0) {
                try {
                    Thread.sleep(msPerLoad);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

            String value = store.get(key);
            if (value != null) {
                result.put(key, value);
            }
        }

        countLoadedKeys.addAndGet(keys.size());
        return result;
    }

    @Override
    public Set<String> loadAllKeys() {
        Set<String> result = new HashSet<>(store.keySet());
        List<String> resultList = new ArrayList<>(result);
        Collections.sort(resultList);
        return result;
    }

    @Override
    public void store(String key, String value) {
        store.put(key, value);
    }

    @Override
    public void storeAll(Map<String, String> map) {
        store.putAll(map);
    }

    @Override
    public void delete(String key) {
        store.remove(key);
    }

    @Override
    public void deleteAll(Collection<String> keys) {
        List<String> keysList = new ArrayList<>(keys);
        Collections.sort(keysList);
        for (String key : keys) {
            store.remove(key);
        }
    }

}

```

Could you please check this?
Is there a workaround to avoid this issue in 3.8.5?

Thanks,
Ruxandra

Hi @vbekiaris ,
thanks for the proposed workaround. We usually recommend the "hazelcast.initial.min.cluster.size" property to our clients. However, often they are not able to use it in real-life installations, because they would no longer be able to start the system if for some reason they don't have all the servers available.

Ruxandra and me have continued to debug this issue as well, but unfortunately we also did not found the root cause yet. The stack trace we see when the call to getMap() blocks and we stop the thread is the following:
```java
169343 [pool-8-thread-1] WARN com.hazelcast.spi.ProxyService  - [172.20.0.1]:5701 [dev] [3.8.4] Error while initializing proxy: IMap{name='map4TestMapStore33'}
com.hazelcast.core.HazelcastException: java.lang.InterruptedException: sleep interrupted
	at com.hazelcast.util.ExceptionUtil.peel(ExceptionUtil.java:94)
	at com.hazelcast.util.ExceptionUtil.peel(ExceptionUtil.java:56)
	at com.hazelcast.util.ExceptionUtil.peel(ExceptionUtil.java:52)
	at com.hazelcast.util.ExceptionUtil.rethrow(ExceptionUtil.java:105)
	at com.hazelcast.map.impl.proxy.MapProxySupport.waitUntilLoaded(MapProxySupport.java:591)
	at com.hazelcast.map.impl.proxy.MapProxyImpl.waitUntilLoaded(MapProxyImpl.java:102)
	at com.hazelcast.map.impl.proxy.MapProxySupport.initializeMapStoreLoad(MapProxySupport.java:222)
	at com.hazelcast.map.impl.proxy.MapProxySupport.initialize(MapProxySupport.java:214)
	at com.hazelcast.map.impl.proxy.MapProxyImpl.initialize(MapProxyImpl.java:102)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.doCreateProxy(ProxyRegistry.java:194)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.createProxy(ProxyRegistry.java:184)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.getOrCreateProxyFuture(ProxyRegistry.java:154)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.getOrCreateProxy(ProxyRegistry.java:135)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.getDistributedObject(ProxyServiceImpl.java:147)
	at com.hazelcast.instance.HazelcastInstanceImpl.getDistributedObject(HazelcastInstanceImpl.java:376)
	at com.hazelcast.instance.HazelcastInstanceImpl.getMap(HazelcastInstanceImpl.java:182)
	at com.hazelcast.instance.HazelcastInstanceProxy.getMap(HazelcastInstanceProxy.java:96)
	at com.nm.test.hazelcast.mapstore.TestMapStore33$1.run(TestMapStore33.java:133)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.InterruptedException: sleep interrupted
	at java.lang.Thread.sleep(Native Method)
	at java.lang.Thread.sleep(Thread.java:340)
	at java.util.concurrent.TimeUnit.sleep(TimeUnit.java:360)
	at com.hazelcast.map.impl.proxy.MapProxySupport.waitAllTrue(MapProxySupport.java:611)
	at com.hazelcast.map.impl.proxy.MapProxySupport.waitUntilLoaded(MapProxySupport.java:589)
	... 18 more
```
maybe this helps...

thanks for looking into this and best,
Lukas

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/operation/TriggerLoadIfNeededOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/operation/TriggerLoadIfNeededOperation.java
@@ -20,6 +20,9 @@ import com.hazelcast.map.impl.MapDataSerializerHook;
 import com.hazelcast.spi.PartitionAwareOperation;
 import com.hazelcast.spi.ReadonlyOperation;
 
+/**
+ * Triggers key loading on member with SENDER or SENDER_BACKUP key loader role, if keys have not finished loading yet.
+ */
 public class TriggerLoadIfNeededOperation extends MapOperation implements PartitionAwareOperation, ReadonlyOperation {
 
     private Boolean isLoaded;
@@ -33,7 +36,7 @@ public class TriggerLoadIfNeededOperation extends MapOperation implements Partit
 
     @Override
     public void run() {
-        isLoaded = recordStore.isLoaded();
+        isLoaded = recordStore.isKeyLoadFinished();
         recordStore.maybeDoInitialLoad();
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/BasicRecordStoreLoader.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/BasicRecordStoreLoader.java
@@ -29,7 +29,6 @@ import com.hazelcast.spi.ExecutionService;
 import com.hazelcast.spi.NodeEngine;
 import com.hazelcast.spi.Operation;
 import com.hazelcast.spi.OperationAccessor;
-import com.hazelcast.spi.OperationResponseHandler;
 import com.hazelcast.spi.OperationService;
 import com.hazelcast.spi.properties.GroupProperty;
 import com.hazelcast.util.ExceptionUtil;
@@ -44,7 +43,6 @@ import java.util.Map;
 import java.util.Queue;
 import java.util.concurrent.Callable;
 import java.util.concurrent.Future;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import static com.hazelcast.spi.ExecutionService.MAP_LOADER_EXECUTOR;
@@ -54,8 +52,6 @@ import static com.hazelcast.spi.ExecutionService.MAP_LOADER_EXECUTOR;
  */
 class BasicRecordStoreLoader implements RecordStoreLoader {
 
-    private final RecordStore recordStore;
-    private final AtomicBoolean loaded;
     private final ILogger logger;
     private final String name;
     private final MapServiceContext mapServiceContext;
@@ -64,13 +60,11 @@ class BasicRecordStoreLoader implements RecordStoreLoader {
 
     BasicRecordStoreLoader(RecordStore recordStore) {
         final MapContainer mapContainer = recordStore.getMapContainer();
-        this.recordStore = recordStore;
         this.name = mapContainer.getName();
         this.mapServiceContext = mapContainer.getMapServiceContext();
         this.partitionId = recordStore.getPartitionId();
         this.mapDataStore = recordStore.getMapDataStore();
         this.logger = mapServiceContext.getNodeEngine().getLogger(getClass());
-        this.loaded = new AtomicBoolean(false);
     }
 
     @Override
@@ -118,7 +112,6 @@ class BasicRecordStoreLoader implements RecordStoreLoader {
         removeUnloadableKeys(keys);
 
         if (keys.isEmpty()) {
-            loaded.set(true);
             return;
         }
 
@@ -144,9 +137,6 @@ class BasicRecordStoreLoader implements RecordStoreLoader {
             final List<Data> chunk = batchChunks.poll();
             final List<Data> keyValueSequence = loadAndGet(chunk);
             if (keyValueSequence.isEmpty()) {
-                if (finishedBatchCounter.decrementAndGet() == 0) {
-                    loaded.set(true);
-                }
                 continue;
             }
             futures.add(sendOperation(keyValueSequence, finishedBatchCounter));
@@ -225,14 +215,6 @@ class BasicRecordStoreLoader implements RecordStoreLoader {
         MapOperationProvider operationProvider = mapServiceContext.getMapOperationProvider(name);
         MapOperation operation = operationProvider.createPutFromLoadAllOperation(name, keyValueSequence);
         operation.setNodeEngine(nodeEngine);
-        operation.setOperationResponseHandler(new OperationResponseHandler() {
-            @Override
-            public void sendResponse(Operation op, Object obj) {
-                if (finishedBatchCounter.decrementAndGet() == 0) {
-                    loaded.set(true);
-                }
-            }
-        });
         operation.setPartitionId(partitionId);
         OperationAccessor.setCallerAddress(operation, nodeEngine.getThisAddress());
         operation.setCallerUuid(nodeEngine.getLocalMember().getUuid());

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapStoreDataLoadingContinuesWhenNodeJoins.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapStoreDataLoadingContinuesWhenNodeJoins.java
@@ -27,16 +27,23 @@ import com.hazelcast.core.MapStore;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.logging.Logger;
 import com.hazelcast.test.AssertTask;
-import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.HazelcastParametersRunnerFactory;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.SlowTest;
 import com.hazelcast.util.EmptyStatement;
+import org.junit.After;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+import org.junit.runners.Parameterized.UseParametersRunnerFactory;
 
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
@@ -49,7 +56,10 @@ import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReferenceArray;
 
+import static com.hazelcast.config.MapStoreConfig.InitialLoadMode.EAGER;
+import static com.hazelcast.config.MapStoreConfig.InitialLoadMode.LAZY;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.fail;
 
@@ -58,7 +68,8 @@ import static org.junit.Assert.fail;
  * <p/>
  * Thanks Lukas Blunschi (@lukasblu) for contributing this test originally.
  */
-@RunWith(HazelcastParallelClassRunner.class)
+@RunWith(Parameterized.class)
+@UseParametersRunnerFactory(HazelcastParametersRunnerFactory.class)
 @Category(SlowTest.class)
 public class MapStoreDataLoadingContinuesWhenNodeJoins extends HazelcastTestSupport {
 
@@ -67,68 +78,64 @@ public class MapStoreDataLoadingContinuesWhenNodeJoins extends HazelcastTestSupp
     private static final int WRITE_DELAY_SECONDS = 5;
     private static final int PRELOAD_SIZE = 1000;
     private static final int MS_PER_LOAD = 300;
+    private static final int NODE_COUNT = 2;
 
     private static final ILogger LOGGER = Logger.getLogger(MapStoreDataLoadingContinuesWhenNodeJoins.class);
 
-    private static final AtomicLong thread1Shutdown = new AtomicLong(-1L);
-    private static final AtomicLong thread2Shutdown = new AtomicLong(-1L);
-    private static final AtomicInteger mapSize = new AtomicInteger(-1);
+    @Parameter
+    public InitialLoadMode initialLoadMode;
+
+    private final AtomicReferenceArray<HazelcastInstance> instances = new AtomicReferenceArray<HazelcastInstance>(NODE_COUNT);
+
+    @Parameters(name = "{0}")
+    public static Collection<Object[]> parameters() {
+        return Arrays.asList(new Object[][]{
+                {LAZY},
+                {EAGER}
+        });
+    }
+
+    @After
+    public void tearDown() {
+        for (int i = 0; i < NODE_COUNT; i++) {
+            try {
+                instances.get(i).getLifecycleService().terminate();
+            } catch (Throwable t) {
+                EmptyStatement.ignore(t);
+            }
+        }
+    }
 
     @Test(timeout = 600000)
     public void testNoDeadLockDuringJoin() throws Exception {
         // create shared hazelcast config
-        final Config config = new XmlConfigBuilder().build();
-
-        // disable JMX to make sure lazy loading works asynchronously
-        config.setProperty("hazelcast.jmx", "false");
-        // get map config
-        MapConfig mapConfig = config.getMapConfig(MAP_NAME);
-
-        // create shared map store implementation
-        // - use slow loading (300ms per map entry)
-        final CountDownLatch halfOfKeysAreLoaded = new CountDownLatch(1);
-        final InMemoryMapStore store = new InMemoryMapStore(halfOfKeysAreLoaded, MS_PER_LOAD, false);
-        store.preload(PRELOAD_SIZE);
-
-        // configure map store
-        MapStoreConfig mapStoreConfig = new MapStoreConfig();
-        mapStoreConfig.setEnabled(true);
-        mapStoreConfig.setInitialLoadMode(InitialLoadMode.LAZY);
-        mapStoreConfig.setWriteDelaySeconds(WRITE_DELAY_SECONDS);
-        mapStoreConfig.setClassName(null);
-        mapStoreConfig.setImplementation(store);
-        mapConfig.setMapStoreConfig(mapStoreConfig);
-        final TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(2);
+        final Config config = createConfigWithDelayingMapStore();
+        final TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(NODE_COUNT);
 
         final CountDownLatch node1Started = new CountDownLatch(1);
         final CountDownLatch node1FinishedLoading = new CountDownLatch(1);
-        // thread 1:
-        // start a single node and load the data
+        final AtomicLong thread1Finished = new AtomicLong(-1L);
+        final AtomicLong thread2Finished = new AtomicLong(-1L);
+
+        // thread 1: start a single member and load the data
         Thread thread1 = new Thread(new Runnable() {
 
             @Override
             public void run() {
-
                 HazelcastInstance hcInstance = factory.newHazelcastInstance(config);
-                // try-finally to stop hazelcast instance
+                instances.set(0, hcInstance);
                 node1Started.countDown();
-                try {
-                    // get map
-                    // this will trigger loading the data
-                    IMap<String, String> map = hcInstance.getMap(MAP_NAME);
-                    map.size();
-                    node1FinishedLoading.countDown();
-                } finally {
-                    thread1Shutdown.set(System.currentTimeMillis());
-                    hcInstance.getLifecycleService().shutdown();
-                }
+                // get map and trigger loading the data
+                IMap<String, String> map = hcInstance.getMap(MAP_NAME);
+                map.size();
+                node1FinishedLoading.countDown();
+                thread1Finished.set(System.currentTimeMillis());
             }
         }, "Thread 1");
         thread1.start();
 
         node1Started.await();
-        // thread 2:
-        // simulate a second member which joins the cluster
+        // thread 2: a second member joins the cluster
         Thread thread2 = new Thread(new Runnable() {
             @Override
             public void run() {
@@ -137,11 +144,9 @@ public class MapStoreDataLoadingContinuesWhenNodeJoins extends HazelcastTestSupp
                     hcInstance.getMap(MAP_NAME);
                     final int loadTimeMillis = MS_PER_LOAD * PRELOAD_SIZE;
                     node1FinishedLoading.await(loadTimeMillis, TimeUnit.MILLISECONDS);
+                    thread2Finished.set(System.currentTimeMillis());
                 } catch (InterruptedException e) {
                     EmptyStatement.ignore(e);
-                } finally {
-                    thread2Shutdown.set(System.currentTimeMillis());
-                    hcInstance.getLifecycleService().shutdown();
                 }
             }
         }, "Thread 2");
@@ -152,35 +157,72 @@ public class MapStoreDataLoadingContinuesWhenNodeJoins extends HazelcastTestSupp
         thread2.join();
 
         // assert correct shutdown order
-        if (thread1Shutdown.get() > thread2Shutdown.get()) {
+        if (thread1Finished.get() > thread2Finished.get()) {
             fail("Thread 2 was shutdown before thread 1.");
         }
     }
 
     @Test(timeout = 600000)
+    public void testLoadingFinishes_whenMemberJoinsWhileLoading() throws Exception {
+        final Config config = createConfigWithDelayingMapStore();
+
+        final TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(2);
+
+        final CountDownLatch node1MapLoadingAboutToStart = new CountDownLatch(1);
+        final CountDownLatch node1FinishedLoading = new CountDownLatch(1);
+        final AtomicInteger mapSizeOnNode2 = new AtomicInteger();
+
+        // thread 1: start a single node and trigger loading the data
+        Thread thread1 = new Thread(new Runnable() {
+
+            @Override
+            public void run() {
+                HazelcastInstance instance = factory.newHazelcastInstance(config);
+                instances.set(0, instance);
+                // get map and trigger loading the data
+                IMap<String, String> map = instance.getMap(MAP_NAME);
+                node1MapLoadingAboutToStart.countDown();
+                map.size();
+                node1FinishedLoading.countDown();
+            }
+        }, "Thread 1");
+        thread1.start();
+
+        node1MapLoadingAboutToStart.await();
+        // thread 2: second member joins the cluster while loading is in progress
+        Thread thread2 = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                HazelcastInstance instance = factory.newHazelcastInstance(config);
+                instances.set(1, instance);
+                try {
+                    IMap map = instance.getMap(MAP_NAME);
+                    final int loadTimeMillis = MS_PER_LOAD * PRELOAD_SIZE;
+                    node1FinishedLoading.await(loadTimeMillis, TimeUnit.MILLISECONDS);
+                    mapSizeOnNode2.set(map.size());
+                } catch (InterruptedException e) {
+                    EmptyStatement.ignore(e);
+                }
+            }
+        }, "Thread 2");
+        thread2.start();
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                assertEquals(PRELOAD_SIZE, mapSizeOnNode2.get());
+            }
+        });
+
+        // join threads
+        thread1.join();
+        thread2.join();
+    }
+
+    @Test(timeout = 600000)
     public void testDataLoadedCorrectly() throws Exception {
         // create shared hazelcast config
-        final Config config = new XmlConfigBuilder().build();
-        // disable JMX to make sure lazy loading works asynchronously
-        config.setProperty("hazelcast.jmx", "false");
-
-        // get map config
-        MapConfig mapConfig = config.getMapConfig(MAP_NAME);
-
-        // create shared map store implementation
-        // - use slow loading (300ms per map entry)
-        final CountDownLatch halfOfKeysAreLoaded = new CountDownLatch(1);
-        final InMemoryMapStore store = new InMemoryMapStore(halfOfKeysAreLoaded, 300, false);
-        store.preload(PRELOAD_SIZE);
-
-        // configure map store
-        MapStoreConfig mapStoreConfig = new MapStoreConfig();
-        mapStoreConfig.setEnabled(true);
-        mapStoreConfig.setInitialLoadMode(InitialLoadMode.LAZY);
-        mapStoreConfig.setWriteDelaySeconds(WRITE_DELAY_SECONDS);
-        mapStoreConfig.setClassName(null);
-        mapStoreConfig.setImplementation(store);
-        mapConfig.setMapStoreConfig(mapStoreConfig);
+        final Config config = createConfigWithDelayingMapStore();
 
         final TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(2);
 
@@ -194,26 +236,20 @@ public class MapStoreDataLoadingContinuesWhenNodeJoins extends HazelcastTestSupp
             @Override
             public void run() {
                 HazelcastInstance instance = factory.newHazelcastInstance(config);
-                // try-finally to stop hazelcast instance
+                instances.set(0, instance);
                 node1Started.countDown();
-                try {
-
-                    // get map
-                    // this will trigger loading the data
-                    final IMap<String, String> map = instance.getMap(MAP_NAME);
-                    mapSize.set(map.size());
-                    node1FinishedLoading.countDown();
-                    assertTrueEventually(new AssertTask() {
-                        @Override
-                        public void run() {
-                            assertEquals(PRELOAD_SIZE, map.size());
-                        }
-                    }, 5);
-                    // -------------------------------------------------- {20s}
-
-                } finally {
-                    instance.getLifecycleService().shutdown();
-                }
+                // get map
+                // this will trigger loading the data
+                final IMap<String, String> map = instance.getMap(MAP_NAME);
+                map.size();
+                node1FinishedLoading.countDown();
+                assertTrueEventually(new AssertTask() {
+                    @Override
+                    public void run() throws Exception {
+                        assertEquals(PRELOAD_SIZE, map.size());
+                    }
+                }, 5);
+                // -------------------------------------------------- {20s}
             }
         }, "Thread 1");
         thread1.start();
@@ -226,7 +262,7 @@ public class MapStoreDataLoadingContinuesWhenNodeJoins extends HazelcastTestSupp
             @Override
             public void run() {
                 HazelcastInstance instance = factory.newHazelcastInstance(config);
-                // try-finally to stop hazelcast instance
+                instances.set(1, instance);
                 try {
                     // get map
                     instance.getMap(MAP_NAME);
@@ -234,8 +270,6 @@ public class MapStoreDataLoadingContinuesWhenNodeJoins extends HazelcastTestSupp
                     node1FinishedLoading.await(loadTimeMillis, TimeUnit.MILLISECONDS);
                 } catch (InterruptedException e) {
                     EmptyStatement.ignore(e);
-                } finally {
-                    instance.getLifecycleService().shutdown();
                 }
             }
         }, "Thread 2");
@@ -346,5 +380,31 @@ public class MapStoreDataLoadingContinuesWhenNodeJoins extends HazelcastTestSupp
         if (log) {
             LOGGER.info("Slept " + TimeUnit.MILLISECONDS.toSeconds(ms) + "seconds.");
         }
+    }
+
+    private Config createConfigWithDelayingMapStore() {
+        // create shared hazelcast config
+        final Config config = new XmlConfigBuilder().build();
+
+        // disable JMX to make sure lazy loading works asynchronously
+        config.setProperty("hazelcast.jmx", "false");
+        // get map config
+        MapConfig mapConfig = config.getMapConfig(MAP_NAME);
+
+        // create shared map store implementation
+        // - use slow loading (300ms per map entry)
+        final CountDownLatch halfOfKeysAreLoaded = new CountDownLatch(1);
+        final InMemoryMapStore store = new InMemoryMapStore(halfOfKeysAreLoaded, MS_PER_LOAD, false);
+        store.preload(PRELOAD_SIZE);
+
+        // configure map store
+        MapStoreConfig mapStoreConfig = new MapStoreConfig();
+        mapStoreConfig.setEnabled(true);
+        mapStoreConfig.setInitialLoadMode(initialLoadMode);
+        mapStoreConfig.setWriteDelaySeconds(WRITE_DELAY_SECONDS);
+        mapStoreConfig.setClassName(null);
+        mapStoreConfig.setImplementation(store);
+        mapConfig.setMapStoreConfig(mapStoreConfig);
+        return config;
     }
 }

